securesignals-ima-devapp/src/main/java/com/uid2/dev/MainActivity.java

```diff
@@ -0,0 +1,283 @@
+                                        // AdEventType.CONTENT_RESUME_REQUESTED is fired when the ad
+                                        // you should play your content.
+                                        resumeContent();
+                                        break;
+                                    case ALL_ADS_COMPLETED:
+                                        // Calling adsManager.destroy() triggers the function
+                                        // VideoAdPlayer.release().
+                                        adsManager.destroy();
+                                        adsManager = null;
+                                        break;
+                                    case CLICKED:
+                                        // When the user clicks on the Learn More button, the IMA SDK fires
+                                        // this event, pauses the ad, and opens the ad's click-through URL.
+                                        // When the user returns to the app, the IMA SDK calls the
+                                        // VideoAdPlayer.playAd() function automatically.
+                                        break;
+                                    default:
+                                        break;
+                                }
+                            }
+                        });
+                    AdsRenderingSettings adsRenderingSettings =
+                        ImaSdkFactory.getInstance().createAdsRenderingSettings();
+                    adsManager.init(adsRenderingSettings);
+                }
+            });
+
+        // When the play button is clicked, request ads and hide the button.
+        playButton = findViewById(R.id.playButton);
+        playButton.setOnClickListener(
+            view -> {
+                videoPlayer.setVideoPath(SAMPLE_VIDEO_URL);
+                requestAds(SAMPLE_VAST_TAG_URL);
+                view.setVisibility(View.GONE);
+            });
+    }
+
+    private void loadUID2Identity() {
+
+        InputStream is = getResources().openRawResource(R.raw.uid2identity);
+        StringBuilder text = new StringBuilder();
+
+        try (BufferedReader br = new BufferedReader(new InputStreamReader(is))) {
+            String line;
+
+            while ((line = br.readLine()) != null) {
+                text.append(line);
+                text.append('\n');
+            }
+
+            String jsonString = text.toString();
+
+            JSONObject jsonObject = (JSONObject) new JSONTokener(jsonString).nextValue();
+
+            // Emulate A UID2Identity With Valid Times
+            long identityExpires = Instant.now().toEpochMilli() * 60 * 60;
+            long refreshFrom = Instant.now().toEpochMilli() * 60 * 40;
+            long refreshExpires = Instant.now().toEpochMilli() * 60 * 50;
+
+            UID2Identity identity = new UID2Identity(jsonObject.getString("advertising_token"),
```

Comment:
The `UID2Identity` already has a way to deserialise a json representation. However, because you're using Java, it isn't accessible (good spot!). We should make the following change to the SDK:

```
diff --git a/sdk/src/main/java/com/uid2/data/UID2Identity.kt b/sdk/src/main/java/com/uid2/data/UID2Identity.kt
index ad94221..b4444de 100644
--- a/sdk/src/main/java/com/uid2/data/UID2Identity.kt
+++ b/sdk/src/main/java/com/uid2/data/UID2Identity.kt
@@ -34,6 +34,7 @@ data class UID2Identity(
          * Helper function to parse a given JSON object into the expected UID2Identity instance. If the JSON instance
          * doesn't contain all required parameters, then null is returned.
          */
+        @JvmStatic
         fun fromJson(json: JSONObject): UID2Identity? {
             val advertisingToken = json.opt("advertising_token")?.toString() ?: return null
             val refreshToken = json.opt("refresh_token")?.toString() ?: return null
```

This could then be `UID2Identity.Companion.fromJson`. We would have to then mutate the times, by creating a new identity instance, but I think that's cleaner than duplicating the deserialisation logic.

Reply:
Done.